File: api/cost_structure_details.py
"""Phase 4 cost-structure import, fee categories, and budget item cost properties.

Legacy anchors: CostStructureImport.cs and FormBudgetCostProperty.cs.
"""
from __future__ import annotations

import json
from datetime import datetime, timezone
from uuid import uuid4

from flask import Blueprint, jsonify, request
from sqlalchemy import Boolean, Column, DateTime, Integer, MetaData, Numeric, String, Table, Text, and_, delete, select

metadata = MetaData()

cost_structure_categories = Table(
    "cost_structure_categories", metadata,
    Column("id", String(100), primary_key=True),
    Column("cost_structure_type_id", String(100), nullable=False, index=True),
    Column("code", String(100), nullable=False),
    Column("name", String(300), nullable=False),
    Column("kind", String(30), nullable=False),
    Column("sequence", Integer, nullable=False),
    Column("rate", Numeric(28, 8), nullable=False),
    Column("enabled", Boolean, nullable=False, default=True),
    Column("row_version", Integer, nullable=False, default=1),
)

budget_item_cost_properties = Table(
    "budget_item_cost_properties", metadata,
    Column("project_code", String(100), primary_key=True),
    Column("budget_item_id", String(100), primary_key=True),
    Column("cost_category_id", String(100), nullable=False, index=True),
    Column("cost_kind", String(30), nullable=False),
    Column("sign", Integer, nullable=False, default=1),
    Column("rate", Numeric(28, 8), nullable=False),
    Column("updated_by", String(100), nullable=False),
    Column("updated_at", DateTime(timezone=True), nullable=False),
    Column("row_version", Integer, nullable=False, default=1),
)

cost_structure_import_runs = Table(
    "cost_structure_import_runs", metadata,
    Column("id", String(100), primary_key=True),
    Column("cost_structure_type_id", String(100), nullable=False, index=True),
    Column("only_structure", Boolean, nullable=False),
    Column("status", String(30), nullable=False),
    Column("total_rows", Integer, nullable=False),
    Column("imported_rows", Integer, nullable=False),
    Column("errors_json", Text, nullable=False),
    Column("created_by", String(100), nullable=False),
    Column("created_at", DateTime(timezone=True), nullable=False),
)

_ALLOWED_KINDS = {"DIRECT", "INDIRECT", "MANAGEMENT", "TAX", "PERCENT", "ADJUSTMENT"}


class CostStructureDetailService:
    def __init__(self, engine):
        self.engine = engine
        metadata.create_all(engine)

    def import_definition(self, type_id: str, payload: dict, actor: str, only_structure: bool = False) -> dict:
        categories = payload.get("categories") or []
        if not isinstance(categories, list) or not categories:
            raise ValueError("categories must be a non-empty list")
        normalized, errors = [], []
        seen = set()
        for index, row in enumerate(categories, start=1):
            try:
                code = str(row.get("code", "")).strip().upper()
                name = str(row.get("name", "")).strip()
                kind = str(row.get("kind", "DIRECT")).strip().upper()
                if not code or not name:
                    raise ValueError("code and name are required")
                if code in seen:
                    raise ValueError("duplicate category code")
                if kind not in _ALLOWED_KINDS:
                    raise ValueError("unsupported cost kind")
                seen.add(code)
                normalized.append({
                    "id": str(row.get("id") or f"{type_id}:{code}"),
                    "cost_structure_type_id": type_id,
                    "code": code,
                    "name": name,
                    "kind": kind,
                    "sequence": int(row.get("sequence", index)),
                    "rate": str(row.get("rate", "0")),
                    "enabled": bool(row.get("enabled", True)),
                    "row_version": 1,
                })
            except Exception as exc:
                errors.append({"row": index, "detail": str(exc)})
        if errors:
            raise ValueError(json.dumps(errors, ensure_ascii=False))
        run_id, now = str(uuid4()), datetime.now(timezone.utc)
        with self.engine.begin() as conn:
            from api.cost_structure import cost_structure_types
            target = conn.execute(select(cost_structure_types.c.id).where(cost_structure_types.c.id == type_id)).first()
            if not target:
                raise LookupError("cost structure type not found")
            conn.execute(delete(cost_structure_categories).where(cost_structure_categories.c.cost_structure_type_id == type_id))
            conn.execute(cost_structure_categories.insert(), normalized)
            conn.execute(cost_structure_import_runs.insert().values(
                id=run_id, cost_structure_type_id=type_id, only_structure=bool(only_structure),
                status="COMPLETED", total_rows=len(normalized), imported_rows=len(normalized),
                errors_json="[]", created_by=actor, created_at=now,
            ))
        return {"id": run_id, "status": "COMPLETED", "cost_structure_type_id": type_id,
                "only_structure": bool(only_structure), "total_rows": len(normalized), "imported_rows": len(normalized)}

    def list_categories(self, type_id: str) -> list[dict]:
        stmt = select(cost_structure_categories).where(
            cost_structure_categories.c.cost_structure_type_id == type_id
        ).order_by(cost_structure_categories.c.sequence, cost_structure_categories.c.code)
        with self.engine.connect() as conn:
            return [dict(row) for row in conn.execute(stmt).mappings().all()]

    def save_item_property(self, project_code: str, item_id: str, body: dict, actor: str) -> dict:
        category_id = str(body.get("cost_category_id", "")).strip()
        cost_kind = str(body.get("cost_kind", "DIRECT")).strip().upper()
        sign = int(body.get("sign", 1))
        if not category_id:
            raise ValueError("cost_category_id is required")
        if cost_kind not in _ALLOWED_KINDS:
            raise ValueError("unsupported cost kind")
        if sign not in {-1, 1}:
            raise ValueError("sign must be -1 or 1")
        now = datetime.now(timezone.utc)
        requested = int(body.get("row_version", 0))
        with self.engine.begin() as conn:
            category = conn.execute(select(cost_structure_categories).where(and_(
                cost_structure_categories.c.id == category_id,
                cost_structure_categories.c.enabled.is_(True),
            ))).mappings().first()
            if not category:
                raise LookupError("enabled cost category not found")
            current = conn.execute(select(budget_item_cost_properties).where(and_(
                budget_item_cost_properties.c.project_code == project_code,
                budget_item_cost_properties.c.budget_item_id == item_id,
            ))).mappings().first()
            values = dict(cost_category_id=category_id, cost_kind=cost_kind, sign=sign,
                          rate=str(body.get("rate", category["rate"])), updated_by=actor, updated_at=now)
            if current:
                if requested != int(current["row_version"]):
                    raise RuntimeError("CONFLICT")
                result = conn.execute(budget_item_cost_properties.update().where(and_(
                    budget_item_cost_properties.c.project_code == project_code,
                    budget_item_cost_properties.c.budget_item_id == item_id,
                    budget_item_cost_properties.c.row_version == requested,
                )).values(**values, row_version=requested + 1))
                if result.rowcount != 1:
                    raise RuntimeError("CONFLICT")
            else:
                if requested != 0:
                    raise RuntimeError("CONFLICT")
                conn.execute(budget_item_cost_properties.insert().values(
                    project_code=project_code, budget_item_id=item_id, row_version=1, **values,
                ))
        return self.get_item_property(project_code, item_id)

    def get_item_property(self, project_code: str, item_id: str) -> dict:
        stmt = select(
            budget_item_cost_properties,
            cost_structure_categories.c.code.label("category_code"),
            cost_structure_categories.c.name.label("category_name"),
        ).join(cost_structure_categories, cost_structure_categories.c.id == budget_item_cost_properties.c.cost_category_id).where(and_(
            budget_item_cost_properties.c.project_code == project_code,
            budget_item_cost_properties.c.budget_item_id == item_id,
        ))
        with self.engine.connect() as conn:
            row = conn.execute(stmt).mappings().first()
        if not row:
            raise LookupError("budget item cost property not found")
        item = dict(row)
        item["rate"] = str(item["rate"])
        item["updated_at"] = item["updated_at"].isoformat()
        item["deep_link"] = f"/app/budget/{project_code}?item={item_id}&panel=cost-property"
        return item


def build_cost_structure_detail_blueprint(service: CostStructureDetailService, resolve_user_id):
    bp = Blueprint("cost_structure_details", __name__, url_prefix="/api/cost-structures")

    def actor() -> str:
        value = resolve_user_id()
        if value is None:
            raise PermissionError("authentication required")
        return str(value)

    @bp.post("/types/<type_id>/import")
    def import_definition(type_id: str):
        body = request.get_json(silent=True) or {}
        try:
            return jsonify(service.import_definition(type_id, body, actor(), bool(body.get("only_structure", False))))
        except PermissionError as exc:
            return jsonify({"code": "UNAUTHORIZED", "detail": str(exc)}), 401
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404
        except ValueError as exc:
            return jsonify({"code": "INVALID_ARGUMENT", "detail": str(exc)}), 400

    @bp.get("/types/<type_id>/categories")
    def list_categories(type_id: str):
        return jsonify(service.list_categories(type_id))

    @bp.get("/projects/<project_code>/items/<item_id>/cost-property")
    def get_property(project_code: str, item_id: str):
        try:
            return jsonify(service.get_item_property(project_code, item_id))
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404

    @bp.put("/projects/<project_code>/items/<item_id>/cost-property")
    def save_property(project_code: str, item_id: str):
        try:
            return jsonify(service.save_item_property(project_code, item_id, request.get_json(silent=True) or {}, actor()))
        except PermissionError as exc:
            return jsonify({"code": "UNAUTHORIZED", "detail": str(exc)}), 401
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404
        except ValueError as exc:
            return jsonify({"code": "INVALID_ARGUMENT", "detail": str(exc)}), 400
        except RuntimeError:
            return jsonify({"code": "CONFLICT", "detail": "stale row_version"}), 409

    return bp

File: pcces-go/internal/storage/sqlite/cost_structure_details.go
package sqlite

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	errx "github.com/liuxb99/pcces-web/pcces-go/internal/platform/errors"
)

var allowedCostKinds = map[string]bool{
	"DIRECT": true, "INDIRECT": true, "MANAGEMENT": true,
	"TAX": true, "PERCENT": true, "ADJUSTMENT": true,
}

type CostStructureCategory struct {
	ID                  string `json:"id"`
	CostStructureTypeID string `json:"cost_structure_type_id"`
	Code                string `json:"code"`
	Name                string `json:"name"`
	Kind                string `json:"kind"`
	Sequence            int    `json:"sequence"`
	Rate                string `json:"rate"`
	Enabled             bool   `json:"enabled"`
	RowVersion          int64  `json:"row_version"`
}

type CostStructureImportRequest struct {
	OnlyStructure bool                    `json:"only_structure"`
	ActorID       string                  `json:"actor_id"`
	Categories    []CostStructureCategory `json:"categories"`
}

type CostStructureImportResult struct {
	ID                  string `json:"id"`
	Status              string `json:"status"`
	CostStructureTypeID string `json:"cost_structure_type_id"`
	OnlyStructure       bool   `json:"only_structure"`
	TotalRows           int    `json:"total_rows"`
	ImportedRows        int    `json:"imported_rows"`
}

type BudgetItemCostProperty struct {
	ProjectCode    string `json:"project_code"`
	BudgetItemID   string `json:"budget_item_id"`
	CostCategoryID string `json:"cost_category_id"`
	CostKind       string `json:"cost_kind"`
	Sign           int    `json:"sign"`
	Rate           string `json:"rate"`
	UpdatedBy      string `json:"updated_by"`
	UpdatedAt      string `json:"updated_at"`
	RowVersion     int64  `json:"row_version"`
	CategoryCode   string `json:"category_code"`
	CategoryName   string `json:"category_name"`
	DeepLink       string `json:"deep_link"`
}

type CostStructureDetailRepository struct{ store *Store }

func NewCostStructureDetailRepository(store *Store) *CostStructureDetailRepository {
	return &CostStructureDetailRepository{store: store}
}

func normalizeCategory(typeID string, item CostStructureCategory, index int) (CostStructureCategory, error) {
	item.Code = strings.ToUpper(strings.TrimSpace(item.Code))
	item.Name = strings.TrimSpace(item.Name)
	item.Kind = strings.ToUpper(strings.TrimSpace(item.Kind))
	if item.Kind == "" {
		item.Kind = "DIRECT"
	}
	if item.Code == "" || item.Name == "" {
		return item, errx.New(errx.CodeInvalidArgument, "category code and name are required", "P4-COST-002")
	}
	if !allowedCostKinds[item.Kind] {
		return item, errx.New(errx.CodeInvalidArgument, "unsupported cost kind", "P4-COST-002")
	}
	if item.ID == "" {
		item.ID = typeID + ":" + item.Code
	}
	if item.Sequence == 0 {
		item.Sequence = index + 1
	}
	if item.Rate == "" {
		item.Rate = "0"
	}
	item.CostStructureTypeID = typeID
	item.RowVersion = 1
	return item, nil
}

func (r *CostStructureDetailRepository) ImportDefinition(ctx context.Context, typeID string, req CostStructureImportRequest) (CostStructureImportResult, error) {
	typeID = strings.TrimSpace(typeID)
	if typeID == "" || len(req.Categories) == 0 {
		return CostStructureImportResult{}, errx.New(errx.CodeInvalidArgument, "type id and categories are required", "P4-COST-002")
	}
	seen := map[string]bool{}
	normalized := make([]CostStructureCategory, 0, len(req.Categories))
	for i, row := range req.Categories {
		item, err := normalizeCategory(typeID, row, i)
		if err != nil {
			return CostStructureImportResult{}, err
		}
		if seen[item.Code] {
			return CostStructureImportResult{}, errx.New(errx.CodeInvalidArgument, "duplicate category code", "P4-COST-002")
		}
		seen[item.Code] = true
		normalized = append(normalized, item)
	}

	tx, err := r.store.db.BeginTx(ctx, nil)
	if err != nil {
		return CostStructureImportResult{}, err
	}
	defer tx.Rollback()
	var exists int
	if err = tx.QueryRowContext(ctx, `SELECT 1 FROM cost_structure_types WHERE id=?`, typeID).Scan(&exists); err == sql.ErrNoRows {
		return CostStructureImportResult{}, errx.New(errx.CodeNotFound, "cost structure type not found", "P4-COST-002")
	} else if err != nil {
		return CostStructureImportResult{}, err
	}
	if _, err = tx.ExecContext(ctx, `DELETE FROM cost_structure_categories WHERE cost_structure_type_id=?`, typeID); err != nil {
		return CostStructureImportResult{}, err
	}
	for _, item := range normalized {
		if _, err = tx.ExecContext(ctx, `INSERT INTO cost_structure_categories(id,cost_structure_type_id,code,name,kind,sequence,rate,enabled,row_version) VALUES(?,?,?,?,?,?,?,?,1)`,
			item.ID, typeID, item.Code, item.Name, item.Kind, item.Sequence, item.Rate, boolInt(item.Enabled)); err != nil {
			return CostStructureImportResult{}, err
		}
	}
	now := time.Now().UTC().Format(time.RFC3339Nano)
	runID := fmt.Sprintf("CSI-%d", time.Now().UTC().UnixNano())
	errorsJSON, _ := json.Marshal([]any{})
	if _, err = tx.ExecContext(ctx, `INSERT INTO cost_structure_import_runs(id,cost_structure_type_id,only_structure,status,total_rows,imported_rows,errors_json,created_by,created_at) VALUES(?,?,?,?,?,?,?,?,?)`,
		runID, typeID, boolInt(req.OnlyStructure), "COMPLETED", len(normalized), len(normalized), string(errorsJSON), req.ActorID, now); err != nil {
		return CostStructureImportResult{}, err
	}
	if err = tx.Commit(); err != nil {
		return CostStructureImportResult{}, err
	}
	return CostStructureImportResult{ID: runID, Status: "COMPLETED", CostStructureTypeID: typeID, OnlyStructure: req.OnlyStructure, TotalRows: len(normalized), ImportedRows: len(normalized)}, nil
}

func (r *CostStructureDetailRepository) ListCategories(ctx context.Context, typeID string) ([]CostStructureCategory, error) {
	rows, err := r.store.db.QueryContext(ctx, `SELECT id,cost_structure_type_id,code,name,kind,sequence,rate,enabled,row_version FROM cost_structure_categories WHERE cost_structure_type_id=? ORDER BY sequence,code`, typeID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	items := []CostStructureCategory{}
	for rows.Next() {
		var item CostStructureCategory
		var enabled int
		if err := rows.Scan(&item.ID, &item.CostStructureTypeID, &item.Code, &item.Name, &item.Kind, &item.Sequence, &item.Rate, &enabled, &item.RowVersion); err != nil {
			return nil, err
		}
		item.Enabled = enabled != 0
		items = append(items, item)
	}
	return items, rows.Err()
}

func (r *CostStructureDetailRepository) SaveItemProperty(ctx context.Context, item BudgetItemCostProperty) (BudgetItemCostProperty, error) {
	item.ProjectCode = strings.TrimSpace(item.ProjectCode)
	item.BudgetItemID = strings.TrimSpace(item.BudgetItemID)
	item.CostCategoryID = strings.TrimSpace(item.CostCategoryID)
	item.CostKind = strings.ToUpper(strings.TrimSpace(item.CostKind))
	if item.ProjectCode == "" || item.BudgetItemID == "" || item.CostCategoryID == "" {
		return item, errx.New(errx.CodeInvalidArgument, "project, item and category are required", "P4-COST-004")
	}
	if !allowedCostKinds[item.CostKind] {
		return item, errx.New(errx.CodeInvalidArgument, "unsupported cost kind", "P4-COST-004")
	}
	if item.Sign != -1 && item.Sign != 1 {
		return item, errx.New(errx.CodeInvalidArgument, "sign must be -1 or 1", "P4-COST-004")
	}
	if item.Rate == "" {
		item.Rate = "0"
	}
	tx, err := r.store.db.BeginTx(ctx, nil)
	if err != nil {
		return item, err
	}
	defer tx.Rollback()
	var enabled int
	if err = tx.QueryRowContext(ctx, `SELECT enabled FROM cost_structure_categories WHERE id=?`, item.CostCategoryID).Scan(&enabled); err == sql.ErrNoRows || enabled == 0 {
		return item, errx.New(errx.CodeNotFound, "enabled cost category not found", "P4-COST-004")
	} else if err != nil {
		return item, err
	}
	var current int64
	err = tx.QueryRowContext(ctx, `SELECT row_version FROM budget_item_cost_properties WHERE project_code=? AND budget_item_id=?`, item.ProjectCode, item.BudgetItemID).Scan(&current)
	now := time.Now().UTC().Format(time.RFC3339Nano)
	if err == sql.ErrNoRows {
		if item.RowVersion != 0 {
			return item, errx.New(errx.CodeConflict, "stale row_version", "P4-COST-004")
		}
		_, err = tx.ExecContext(ctx, `INSERT INTO budget_item_cost_properties(project_code,budget_item_id,cost_category_id,cost_kind,sign,rate,updated_by,updated_at,row_version) VALUES(?,?,?,?,?,?,?,?,1)`, item.ProjectCode, item.BudgetItemID, item.CostCategoryID, item.CostKind, item.Sign, item.Rate, item.UpdatedBy, now)
	} else if err != nil {
		return item, err
	} else {
		if current != item.RowVersion {
			return item, errx.New(errx.CodeConflict, "stale row_version", "P4-COST-004")
		}
		res, execErr := tx.ExecContext(ctx, `UPDATE budget_item_cost_properties SET cost_category_id=?,cost_kind=?,sign=?,rate=?,updated_by=?,updated_at=?,row_version=row_version+1 WHERE project_code=? AND budget_item_id=? AND row_version=?`, item.CostCategoryID, item.CostKind, item.Sign, item.Rate, item.UpdatedBy, now, item.ProjectCode, item.BudgetItemID, item.RowVersion)
		if execErr != nil {
			return item, execErr
		}
		n, _ := res.RowsAffected()
		if n != 1 {
			return item, errx.New(errx.CodeConflict, "stale row_version", "P4-COST-004")
		}
	}
	if err != nil {
		return item, err
	}
	if err = tx.Commit(); err != nil {
		return item, err
	}
	return r.GetItemProperty(ctx, item.ProjectCode, item.BudgetItemID)
}

func (r *CostStructureDetailRepository) GetItemProperty(ctx context.Context, projectCode, itemID string) (BudgetItemCostProperty, error) {
	var item BudgetItemCostProperty
	err := r.store.db.QueryRowContext(ctx, `SELECT p.project_code,p.budget_item_id,p.cost_category_id,p.cost_kind,p.sign,p.rate,p.updated_by,p.updated_at,p.row_version,c.code,c.name FROM budget_item_cost_properties p JOIN cost_structure_categories c ON c.id=p.cost_category_id WHERE p.project_code=? AND p.budget_item_id=?`, projectCode, itemID).Scan(&item.ProjectCode, &item.BudgetItemID, &item.CostCategoryID, &item.CostKind, &item.Sign, &item.Rate, &item.UpdatedBy, &item.UpdatedAt, &item.RowVersion, &item.CategoryCode, &item.CategoryName)
	if err == sql.ErrNoRows {
		return item, errx.New(errx.CodeNotFound, "budget item cost property not found", "P4-COST-004")
	}
	if err != nil {
		return item, err
	}
	item.DeepLink = "/app/budget/" + projectCode + "?item=" + itemID + "&panel=cost-property"
	return item, nil
}
